std/digest/hmac.d

```diff
@@ -0,0 +1,283 @@
+ * the second overload.
+ *
+ * This type conforms to $(XREF digest.digest, isDigest).
+ */
+
+template HMAC(H)
+if(isDigest!H && hasBlockSize!H)
+{
+    alias HMAC = HMAC!(H, H.blockSize);
+}
+
+/// ditto
+
+struct HMAC(H, size_t hashBlockSize)
+if(hashBlockSize % 8 == 0)
+{
+@nogc:
+
+    enum blockSize = hashBlockSize;
+
+    private H digest;
+    private ubyte[blockSize / 8] key;
+
+    /**
+     * Constructs the HMAC digest using the specified secret.
+     */
+
+    this(scope const(ubyte)[] secret)
+    {
+        // if secret is too long, shorten it by computing its hash
+        typeof(digest.finish()) buffer = void;
+        if(secret.length > blockSize / 8)
+        {
+            digest.start();
+            digest.put(secret);
+            buffer = digest.finish();
+            secret = buffer[];
+        }
+
+        // if secret is too short, it will be padded with zeroes
+        // (the key buffer is already zero-initialized)
+        import std.algorithm.mutation : copy;
+        secret.copy(key[]);
+
+        start();
+    }
+
+    ///
+    @safe pure nothrow @nogc unittest
+    {
+        import std.digest.sha, std.digest.hmac;
+        import std.string : representation;
+        auto hmac = HMAC!SHA1("My s3cR3T keY".representation);
+    }
+
+    /**
+     * Reinitializes the digest, making it ready for reuse.
+     *
+     * Note:
+     * The constructor leaves the digest in an initialized state, so that this
+     * method only needs to be called if an unfinished digest is to be reused.
+     *
+     * Returns:
+     * A reference to the digest for convenient chaining.
+     */
+
+    ref start() return
+    {
+        ubyte[blockSize / 8] ipad = void;
+        foreach(immutable i; 0 .. blockSize / 8)
+            ipad[i] = key[i] ^ 0x36;
+
+        digest.start();
+        digest.put(ipad[]);
+
+        return this;
+    }
+
+    ///
+    @safe pure nothrow @nogc unittest
+    {
+        import std.digest.sha, std.digest.hmac;
+        import std.string : representation;
+        ubyte[] data1, data2;
+        auto hmac = HMAC!SHA1("My s3cR3T keY".representation);
+        hmac.put(data1);
+        hmac.start();    // reset digest
+        hmac.put(data2); // start over
+    }
+
+    /**
+     * Feeds a piece of data into the hash computation. This method allows the
+     * type to be used as an $(XREF range, OutputRange).
+     *
+     * Returns:
+     * A reference to the digest for convenient chaining.
+     */
+
+    ref put(in ubyte[] data...) return
+    {
+        digest.put(data);
+        return this;
+    }
+
+    ///
+    @safe pure nothrow @nogc unittest
+    {
+        import std.digest.sha, std.digest.hmac;
+        import std.string : representation;
+        ubyte[] data1, data2;
+        auto hmac = HMAC!SHA1("My s3cR3T keY".representation);
+        hmac.put(data1)
+            .put(data2);
+    }
+
+    /**
+     * Resets the digest and returns the finished hash.
+     */
+
+    auto finish()
```

Comment:
What exactly is this returning? I.e. what is the type? Is it a value, a reference, a range, a GC allocated array, a class?